### PR TITLE
Update SDK to 9.0 preview 4

### PR DESCRIPTION
**Pull request by am11**

### CODE REVIEW

**Review by ilonatommy — COMMENTED** [review]
Other places with pinned versions:
- https://github.com/dotnet/runtime/blob/7284f17efe5f1ad7382f65ef3276ac89d594d3d3/src/mono/wasm/testassets/WasmOnAspNetCore/Shared/Shared.csproj#L14
- https://github.com/dotnet/runtime/blob/7284f17efe5f1ad7382f65ef3276ac89d594d3d3/src/mono/wasm/testassets/WasmOnAspNetCore/BlazorClient/BlazorClient.csproj#L15

**Comment by am11** [comment]
Sure, will update those as well. I was mainly following @MichalStrehovsky's footsteps https://github.com/dotnet/runtime/pull/100945 (which made me think only to grep `9.*preview.3`, find versions on the internets and replace) 😅

File: src/coreclr/tools/aot/ILCompiler.TypeSystem/ILCompiler.TypeSystem.csproj
Comment (am11) [inline src/coreclr/tools/aot/ILCompiler.TypeSystem/ILCompiler.TypeSystem.csproj:32]:
```suggestion
    <PackageReference Include="Microsoft.DiaSymReader" Version="$(MicrosoftDiaSymReaderVersion)" Condition="'$(EnableDiaSymReaderUse)' == 'true'" />
```

File: eng/Version.Details.xml
Comment (am11) [inline eng/Version.Details.xml:287]: @ViktorHofer, maybe it would be nice to display 'Product Version' (which contains the commit hash) in the readme of packages going in dotnet feeds? e.g. https://dev.azure.com/dnceng/public/_artifacts/feed/dotnet9/NuGet/System.Reflection.Metadata/overview/9.0.0-preview.5.24273.1

Current approach is to download the package, unzip and use some tool like exiftool to get the ProductVersion to see the commit hash:

```sh
$ exiftool lib/net9.0/System.Reflection.Metadata.dll -productversion
Product Version                 : 9.0.0-preview.5.24273.1+aabefea9ec7c5717177454cd570e48cc55a535bf
```

Comment (akoeplinger) [reply eng/Version.Details.xml:287]: the dependencies in this file are updated automatically so you shouldn't need to bump them for the preview4 update

Comment (am11) [reply eng/Version.Details.xml:287]: The penultimate commit was giving me these errors:

> error CS0246: The type or namespace name 'AssemblyNameInfo' could not be found (are you missing a using directive or an assembly reference?

these version bumps helped (no idea how/why.. 😁).

Comment (akoeplinger) [reply eng/Version.Details.xml:287]: That seems to be related to https://github.com/dotnet/runtime/pull/102489#issuecomment-2125777492.
I suggest we wait until that PR lands.

**Comment by am11** [comment]
`Build wasi-wasm linux Release WasmBuildTests` is failing with:
```
  Discovered:  Wasi.Build.Tests (found 1 of 11 test case)
  Starting:    Wasi.Build.Tests (parallel test collections = on [2 threads], stop on fail = off)
    Wasi.Build.Tests.ILStripTests.WasmStripILAfterAOT_TestDefaultAndOverride(stripILAfterAOT: "", expectILStripping: True, singleFileBundle: False) [STARTING]
    Wasi.Build.Tests.ILStripTests.WasmStripILAfterAOT_TestDefaultAndOverride(stripILAfterAOT: "", expectILStripping: True, singleFileBundle: False) [FAIL]
       Expected 0 exit code but got 150: /root/helix/work/workitem/e/dotnet-latest/dotnet run --no-silent --no-build -c Release
      Standard Output:
      [Release_fkeubaml_415] You must install or update .NET to run this application.
      [Release_fkeubaml_415] 
      [Release_fkeubaml_415] App: /root/helix/work/workitem/e/dotnet-latest/packs/Microsoft.NET.Runtime.WebAssembly.Wasi.Sdk/9.0.0-ci/WasmAppHost/WasmAppHost.dll
      [Release_fkeubaml_415] Architecture: x64
      [Release_fkeubaml_415] Framework: 'Microsoft.NETCore.App', version '9.0.0-preview.4.24266.19' (x64)
      [Release_fkeubaml_415] .NET location: /root/helix/work/workitem/e/dotnet-latest/
      [Release_fkeubaml_415] 
      [Release_fkeubaml_415] The following frameworks were found:
      [Release_fkeubaml_415]   8.0.7 at [/root/helix/work/workitem/e/dotnet-latest/shared/Microsoft.NETCore.App]
      [Release_fkeubaml_415]   9.0.0-preview.4.24251.3 at [/root/helix/work/workitem/e/dotnet-latest/shared/Microsoft.NETCore.App]
      [Release_fkeubaml_415] 
      [Release_fkeubaml_415] Learn more:
      [Release_fkeubaml_415] https://aka.ms/dotnet/app-launch-failed
      [Release_fkeubaml_415] 
      [Release_fkeubaml_415] To install missing framework, download:
      [Release_fkeubaml_415] https://aka.ms/dotnet-core-applaunch?framework=Microsoft.NETCore.App&framework_version=9.0.0-preview.4.24266.19&arch=x64&rid=linux-x64&os=ubuntu.22.04
      
      Stack Trace:
        /_/src/mono/wasm/Wasm.Build.Tests/Common/CommandResult.cs(44,0): at Wasm.Build.Tests.CommandResult.EnsureExitCode(Int32 expectedExitCode, String messagePrefix, Boolean suppressOutput)
        /_/src/mono/wasm/Wasm.Build.Tests/Common/CommandResult.cs(28,0): at Wasm.Build.Tests.CommandResult.EnsureSuccessful(String messagePrefix, Boolean suppressOutput)
        /_/src/mono/wasi/Wasi.Build.Tests/ILStripTests.cs(52,0): at Wasi.Build.Tests.ILStripTests.WasmStripILAfterAOT_TestDefaultAndOverride(String stripILAfterAOT, Boolean expectILStripping, Boolean singleFileBundle)
```

@ilonatommy, is there anything else which needs to be updated before preview.1 -> preview.4 switch is possible for Wasm.Build.Tests?

File: eng/Version.Details.xml
Comment (ericstj) [reply eng/Version.Details.xml:287]: Yep - what happened was Preview4 was rebuilt with newer file versions than Preview5 packages already referenced by the product.  So this part of the repo that builds against .NET 9 without referencing live bits started to get the references from the P4 targeting pack with less API than the P5 packages.  The fix is to update to newer P5 packages.

This is a pretty rare circumstance that arose due to rebuild of P4 - normally we'll be updating the next preview's packages and be past the previous preview's versions by the time we make it public.  Better fixes here would be 1) make all parts of the product build against live bits 2) encode preview in fileversion.  Not sure this will happen often enough to go and do those things.

Comment (am11) [reply eng/Version.Details.xml:287]: > 1) make all parts of the product build against live bits

This is probably the best option. Sometimes it’s a bit tricky to pull it off due to the subsets inter-dependencies. e.g. ilcompiler from clr subset to depend on libs subset for System.Reflection.Metadata in this case. Not impossible to resolve but it’s not particularly straightforward (considering SRM’s own dependencies on live build of ImmutableCollection, Linq etc.).

Comment (ViktorHofer) [reply eng/Version.Details.xml:287]: > @ViktorHofer, maybe it would be nice to display 'Product Version' (which contains the commit hash) in the readme of packages going in dotnet feeds? e.g.

That information is already embedded in the nuspec, isn't it? Some package managers display the SHA. You can always download the nupkg and open it with NuGet Package Explorer or unzip and look at the nuspec manually.

Comment (am11) [reply eng/Version.Details.xml:287]: > That information is already embedded in the nuspec, isn't it?

Yup nuspec has  `<repository type="git" url="https://github.com/dotnet/runtime" commit="aabefea9ec7c5717177454cd570e48cc55a535bf" />`. The readme file can have the version, I remember seeing them in some readmes (as a last line) in the past, can't find an example. That gives a quick, a-the-glance, reference.

Comment (ViktorHofer) [reply eng/Version.Details.xml:287]: This would help when looking at packages on nuget.org, AzDO packaging or via VS package manager? Sounds more like a potential UI improvement for of those components. I don't think the SHA should go there and I would rather prefer to keep the package readme files static.

Correct, some packages have the SHA in the package description. That was done as the nuspec didn't contain the information back then. That should eventually go away.

Comment (am11) [reply eng/Version.Details.xml:287]: Only AzDo; nuget.org's package explorer parses the package number out of nuspec:

image


Yeah, I just thought an informative line won't hurt if it's doable, but if the intent is to keep the readmes static, understood. 👍

Comment (am11) [reply eng/Version.Details.xml]: Reverted this file and merged main. Thanks for the infos everyone!

**Comment by am11** [comment]
> `Build wasi-wasm linux Release WasmBuildTests` is failing with:

Applied @lewing's https://github.com/dotnet/runtime/pull/102624 as a patch and it fixed the error.  👏 
(lets wait until that one is landed then I'll revert and merge main)

**Comment by am11** [comment]
@lewing looks like https://github.com/dotnet/runtime/issues/100275 is back.

**Comment by am11** [comment]
```
[20:31:01] info: Using threadless Xunit runner
[20:31:01] info: Discovering: System.Security.Cryptography.Tests.dll (method display = ClassAndMethod, method display options = None)
[20:31:04] info: Discovered:  System.Security.Cryptography.Tests.dll (found 1815 of 4463 test cases)
[20:31:04] info: Starting:    System.Security.Cryptography.Tests.dll
[20:31:34] info: Finished:    System.Security.Cryptography.Tests.dll
[20:31:40] info: Unable to evaluate script: tab crashed
[20:31:40] info: Unable to evaluate script: tab crashed
[20:31:40] info: Unable to evaluate script: tab crashed
[20:31:40] info: Unable to evaluate script: tab crashed
[20:31:40] info: Unable to evaluate script: tab crashed
[20:31:40] info: Unable to evaluate script: tab crashed
[20:31:40] info: Unable to evaluate script: tab crashed
[20:31:40] info: Unable to evaluate script: tab crashed
[20:31:40] info: Unable to evaluate script: tab crashed
[20:31:40] info: Unable to evaluate script: tab crashed
[20:31:40] info: Unable to evaluate script: tab crashed
[20:31:40] info: Unable to evaluate script: tab crashed
[20:31:40] info: Unable to evaluate script: tab crashed
[20:31:40] info: Unable to evaluate script: tab crashed
[20:31:40] info: Unable to evaluate script: tab crashed
[20:31:40] info: Unable to evaluate script: tab crashed
[20:31:40] info: Unable to evaluate script: tab crashed
[20:31:40] info: Unable to evaluate script: tab crashed
....
```

@akoeplinger could you rerun this leg so we can find out if it's a persistent problem? In previous run it was green which is why I have this feeling. 🤞

**Review by ViktorHofer — APPROVED** [review]

**Pull request merged**
